### DISCUSSION TITLE
Config Subentries

### DISCUSSION THREAD

**Discussion by balloob** [home-assistant/architecture#1070] [General] [answered]
## Context

There are integrations that have authentication stored in a config entry. Some of these integrations really are building blocks that have to be built on top of:

- OpenAI and Google AI contain authentication tokens.
- MQTT contains the broker connection
- KNX contains the Bus connection

MQTT and KNX have in common that we use the config entry to store the authentication, and the entities are created on top of it. They have to be configured in YAML.

For OpenAI and Google AI we decided to limit the user configuration to a single thing, and merge that into the config entry (in this case the prompt). It has the annoying side effect that a user has to re-enter API key for each OpenAI prompt it wants to use.

There are also notifications that are in dire need of a solution. If you add an SMS gateway, how would you define targets? Right now you need to keep your targets in your scripts. Wouldn’t it be great if you could pre-create targets and those are represented as entities ? The scrape integration actually "solved" this and allows defining the sensors based on the scrape information via a very elaborate options flow.

## Solution

We create a new data type called “config subentry”. It is owned by a config entry. It is set up as part of async_setup_entry. Config subentries are created by config subentry flows and updated by config subentry options flows. This works similarly to how we create/edit config flows.

![CleanShot 2024-03-29 at 10 57 51@2x](https://github.com/home-assistant/architecture/assets/1444314/53866a58-f909-4b53-9dc8-33d56d573941)

Config subentries will have a type such that we can decide what config subentry flows to show where in the UI.

For example, config subentries of type `ai_agent` could be shown on a dedicated page like we do for helpers. It would have a button “Add AI Agent” that opens an “Add integration” like flow, but for picking integration/config entry that supports adding “ai_agent” config subentries via a flow.

![CleanShot 2024-03-29 at 10 58 56@2x](https://github.com/home-assistant/architecture/assets/1444314/fa1fc968-c48c-4de6-868c-07979292c48f)

#### Create MQTT entities from UI
New flow with live preview with what the messages from MQTT will result in.

#### Create conversation entities
Add config subentry flows to Google and OpenAI to allow creating customized conversation entities. This will users to create different entities with different prompts, instead of having both OpenAI API key and prompt in 1 config entry.

#### Notification targets
Create targets for notification integrations, like an SMS gateway or Telegram. 

#### Weather: add different locations
Add different weather stations to a single API key.

#### AI Agents

Example of how different AI providing integrations and their config subentries can provide different prompts and capabilities

![CleanShot 2025-01-14 at 08 18 12@2x](https://github.com/user-attachments/assets/116ce124-fdd9-4823-9101-befefac97bed)

And a UI to manage all config subentries across different integrations could look like this:

![CleanShot 2024-03-29 at 11 13 17@2x](https://github.com/home-assistant/architecture/assets/1444314/21acb747-a1e2-469b-9f77-5b09485dc567)

## Lifecycle

Pre-config subentries, the data hierarchy is like this ([source](https://developers.home-assistant.io/docs/architecture/devices-and-services#entity-data-hierarchy)):

![CleanShot 2025-01-14 at 08 19 10@2x](https://github.com/user-attachments/assets/0f28360a-d819-42be-b95a-fe1a28724b90)

With subentries, we are introducing a new optional block between "Config Entry" and "Device Registry Entry". A device can either be linked directly to a config entry or a config subentry, not both. Deleting, disabling or re-enabling a config subentry will follow the existing logic:

> Delete, disable or re-enable any object and all objects below will be adjusted accordingly.

## Alternatives

An alternative is to get all integrations to put all config subentries inside their config entry. 

The downside of this is that it will not be possible to group config subentries from different integrations together on a single page, like showing a page with all AI agents, weather stations or notification targets.

Having the config subentries as a new concept will also allow us to offer edit buttons for entities if we know the relationship between config subentry and entity is 1:1.

## Note

It is not the goal to put automations and scripts into config subentries. Those are inside YAML as per our ADRs and will be kept there.

**Comment by joostlek** [comment]
If I read this some more examples come to mind, can you validate that these are a logical next step?

1. Integrations with an API key per "application". For example Trafikverket_train or Nederlandse spoorwegen. They have 1 API key, but theoretically could use multiple routes. So the API key is the main config entry, and every train/route you want to add is a config subentry. This would also be applicable for integrations like WAQI or weather providers.
2. Integrations with devices that take more information to set up. I think myuplink could make use of this. Recently I have been looking into meross, and that has a part cloud/part local approach, that would also benefit from this as the cloud part could be the config entry (to get device info and controls) and the sub entry to define where this device is located in the local network.

**Reply by joostlek** [reply]
I am also thinking about added value for bringing an integration like `bayesian` to the config flow (since that's one binary sensor that's using a lot of factors with a lot of attributes)

**Reply by balloob** [reply]
I think 1. very much yes. It's about splitting up the authentication from the configuration, if multiple configurations can exist.

Number 2 I might not fully understand. If it's just asking for more information, it could be an options flow.

**Comment by elupus** [comment]
If this also allow discovery, rfxtrx could use this and get rid of its annoying/complex config flow.

**Reply by balloob** [reply]
Didn't think about discovery but that would be an option for sure.

**Comment by janiversen** [comment]
This is a very nice idea, and actually very much like what I have been "toying" with for modbus, at least if I understand it correctly.

Thinking in terms of modbus, there would be a subentry for each transport type, with the specific transport parameters. So in the config entry the user select transport type, and can then set the parameters in the subentry.

The idea that the subentry have a type, is important, because it helps have different subentries.

**Reply by balloob** [reply]
Not sure what "transport type" means here, but I could see a future where one could add a modbus entity under a modbus communication type using a config subentry.

**Reply by janiversen** [reply]
tranport type is "serial", "tcp", "tls" and "udp" (the different ways modbus equipment can be attached to HA) each with a unique set of comm. params.

Using subentries for entries, would demand that one config_entry can hold a list of subentries ! But that would solve the biggest problem I see at the moment.

Being able to "offer" the modbus config_entry as subentry to other integrations, would help overcome some of the configuration issues we see today.

**Reply by balloob** [reply]
A config entry can hold many config subentries. That's the idea :) 

Integration: modbus
Config entry: transport (serial, tcp, tls, udp)
Config subentries: 1 subentry per entity

**Comment by joostlek** [comment]
I think we also need to think about a way to define the flow for the integration. For example, MQTT would maybe be sufficient to only have set up the main config entry, and quit the flow, while another integration might want to register an x amount of flows directly at creation which the user should finish before it can load (To take MyUplink for example, they have devices that can expose up to 600 entities which are not all relevant (but they can't create a generic way to differentiate on that). So in that case it would be nice that the user first has to finish the device flow before the device is added (so the user had to select their relevant data points before adding the device).

Maybe some sort of way to allow the integration or the config flow to "discover" new subentries the user has to configure, that'd be great

**Reply by balloob** [reply]
I don't think we should use this as a way to ask the user more information that is necessary for set up. 600 entities is insane and very much an edge case. 

I do believe that setting up a config entry should create config subentries. For example, configuring OpenAI should create a config subentry for providing an agent for the conversation integration.

**Reply by janiversen** [reply]
What way would you suggest we follow, to ask the user for more information.

Think of the following scenario (modbus, added more text to make it easier understand) ?

Currently I can only do the following:

"Select the way your modbus device is attached (serial, tcp, tls, udp)" ?

and simultaneously show all parameters for each type of attachment (something like 20 in total).

I would like to use something like subentry, to ask the user for more information, like ?

"Select the way your modbus device is attached (serial, tcp, tls, udp)" ?

then show the relevant parameters for the selected way of attachment.

It might be this is not 100% subentry, but having config_flow allowing a hierarchy would make a number of configurations a lot easier and in case of modbus allow a switch to config_flow.

**Reply by Kane610** [reply]
Why can't you just have the next step in the condig flow be based on the previous selection?

**Reply by farmio** [reply]
I would imagine the modbus example to have connection parameters in the config flow:
- menu step for serial, tcp, udp etc. 
- next step detailed infos for the specific type of attachment

and the subentry flow to configure individual entities for that connection 
- menu step for platform 
- next step(s) for addresses, registers etc.

**Reply by janiversen** [reply]
@farmio you expressed my thinking a lot better !

**Comment by Jc2k** [comment]
Could matter use this?

Right now, the integration has a config entry for the connection to the matter addon. With this each pairing would be a sub-config-entry.

Pairings could then be deleted by the normal config entry delete hook.

This could potentially unlock using normal HA discovery flow for adding new matter devices too? AIUI For devices already visible by zeroconf we wouldn't need the app involvement, but previously that didn't really work because the discovery flows end in the creation of config entries.

**Reply by balloob** [reply]
I don't know. In the case of Matter the "pairing" is stored in the Matter SDK, running inside the Matter add-on. The integration is only having representation of that pairing. 

We actually looked at using a normal config flow for Matter pairing and decided against it. It's so complicated that we want to control the UX to bring the optimal experience.

**Comment by elupus** [comment]
Thinking about sub entries... sub entries from another domain could make sense. Say modbus main domain, some heat pump needing to piggy back on the modbus bus config.

**Reply by jbouwh** [reply]
When it is about sub entries from an other domain, it is important we can specialize the schema.
E.g. When we would use `person` sub entries to supply the recipient on a `notify` entity implementation, we should be able to customize the data schema.
So in that case, an integration expose a sub entry feature to be used / linked with additional data to another integration.

For MQTT an alternative method to add manual MQTT items (now it is YAML only) would be a case. Not sure the subentries should be platform subentries from the entity domain, but that could be an idea as well.

**Comment by MartinHjelmare** [comment]
> A device can either be linked directly to a config entry or a config subentry, not both.

It's not clear to me why a device can only be linked to either a config entry or a sub entry. Today it's up to integrations to decide how to link entities to devices/services. Eg one config entry can have single device/service with multiple entities or multiple devices/services with one or many entities. From an integration perspective, adding configuration via config subentries doesn't mean that an integration has a specific relationship between devices/services and its entities. It means that the integration needs to be able to add entities with their own configuration.

I think we should be very careful before we start deciding the relationship between devices/services and entities on behalf of integrations if that's not really needed.

**Reply by balloob** [reply]
I meant with "linked" is about a single relationship. A device can have multiple relationships with a combination of config entries and config subentries. This is all driven by the entities:

Today a device is created by adding an entity to Home Assistant. That entity is tied to either a config subentry or a config entry. That is the one that owns that entity. So if the config subentry is removed, it will remove the config subentry entities, and if the device has no other relationships, the device. The same for if a config entry is removed today.

**Reply by MartinHjelmare** [reply]
Does this mean that entities coming from different config subentries of the same config entry can be linked to the same device as long as that device isn't linked to those different config subentries but to the config entry?

**Reply by emontnemery** [reply]
It means that a device can have multiple subentry links for each config entry it's linked to.
So for example, the links could be:
```
(
    (config entry 1, no subentry),
    (config entry 1, subentry 1),
    (config entry 1, subentry 2),
    (config entry 2, subentry 1),
    (config entry 3, no subentry)
)
```
If one of those links are removed, the entities corresponding to that link are removed.
Links are automatically removed if a config entry or subentry is removed, and can also be manually removed by integrations.

**Reply by MartinHjelmare** [reply]
Thanks!

**Comment by MartinHjelmare** [comment, marked as answer]
This proposal has been approved by the core meeting.

**Reply by frenck** [reply]
They have been implemented now.

https://developers.home-assistant.io/blog/2025/02/16/config-subentries/

**Comment by Martreides** [comment]
Apologies for posting to this discussion if it is not the correct place to do so.

I am currently working on a major overhaul of an integration and am implementing Config Subentries. Only after quite some work I found that Config Subentries is a new functionality and that the integration I am working on is actually mentioned in this thread: nederlandse_spoorwegen.

When designing the Config Subentries were the following two topics taken into account?
1. Accessing data of the config entry in the config subentry while in a config flow. For example to use the API key entered as part of the config entry to test input in the config subentry.
2. The possibility to use the "import step" in a config flow to create config entries and config subentries.

I have opened a draft PR with my implementation of config subentries in the integration: https://github.com/home-assistant/core/pull/139857.

**Reply by scabrero** [reply]
Hi @Martreides, I faced the same problem and this is how I did for point 1:

```
class FooSubentryFlowHandler(ConfigSubentryFlow):

    def _parent_config_entry(self) -> ConfigEntry:
        if self.hass is None:
            raise ValueError("The config entry is not available during initialisation")
        if self.handler is None:
            raise ValueError(
                "The config entry id is not available during initialisation"
            )
        return self.hass.config_entries.async_get_known_entry(self.handler[0])

    async def async_step_foo(self, user_input: dict[str, Any] | None = None) -> SubentryFlowResult:
        config_entry = self._parent_config_entry()
        coordinator: FooDataUpdateCoordinator = config_entry.runtime_data
        api = coordinator.api
```

**Reply by Martreides** [reply]
That seems to work! Thanks a lot for the help!

**Reply by emontnemery** [reply]
This is not the place to ask how to use the API, and I'll mark the discussion as off-topic.

Having said that, feel free to reach out on Discord if you need some help with using the subentry API 👍

**Comment by andystewart999** [comment]
Is it possible to create two sub-entries from a single Config Flow/Subflow?

My example use case is a train travel integration that takes source and dest and returns the next train to make that journey.  It would be nice to offer the user the option of creating the reverse journey as well, and while I could do that with sensors in the same Entry it would also make sense to have a second Entry/SubEntry for the reverse journey - as that's what the result would be if the user created the two journeys separately.

Looking at async_create_entry in the core code it doesn't look like it would support a list of ConfigEntries coming in...

**Reply by gjohansson-ST** [reply]
If you want to discuss such, please take those at HA's discord developer channel. This is not the proper place for how to use things.
